Add support for adding projects concurrently in DependencyGraphSpec using builtIn concurrent type

erdembayar opened the pull request:
## Bug

Fixes: [Nuget/Home#9002](https://github.com/NuGet/Home/issues/9002)

Regression: No  

## Fix

Details: 
The [DependencyGraphSpec ](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.ProjectModel/DependencyGraphSpec.cs#L241) object uses a standard collection object as a backing collection. This makes it impossible to use it in a parallel loop or multiple threads. It should use a concurrent collection instead. 
I have 2 competing idea for this task.
Here is 2nd one which doesn't use synchronization lock. For one which uses synchronization lock please check [PR#3594](https://github.com/NuGet/NuGet.Client/pull/3594).
There are no native framework concurrent SortedSet nor concurrent SortedDictionary, I just decided use ConcurrentDictionary for both of them. Then we sort it before consume it.
Also using ConcurrentDictionary in place of SortedSet may be extra space overhead since we only care about dictionary key not values. Please note ConcurrentBag not suitable for this because it doesn't check for duplicate item and checking it costs O(n) time complexity.
Overall I don't believe replacing SortedSet by ConcurrentDictionary degrade performance, because it's pure C# code usually doesn't tax heavy for small/medium size sets.
Here insert time O(logN) for SortedSet but much cheaper O(1) for Dictionary.
Iterate through values then O(n) for SortedSet but expensive O(n* logN)for Dictionary before consumption.

## Testing/Validation

Tests Added: No  
Reason for not adding tests:  There are already many unit tests covering this code paths.
Validation:  Manual testing.

erdembayar committed:
Replace standard collections with builtin Concurrent collections for multithread safety.
There are no native framework concurrent SortedSet nor concurrent SortedDictionary, I just decided use ConcurrentDictionary for both of them. Then we sort it before consume it.
Also using ConcurrentDictionary in place of SortedSet may be extra space overhead since we only care about dictionary key not values. Please note ConcurrentBag not suitable for this because it doesn't check for duplicate item and checking it costs O(n) time complexity.\
Overall I believe replacing SortedSet by ConcurrentDictionary degrade performance, because it's pure C# code usually doesn't tax heavy for small/medium size sets.
Here insert time O(logN) for SortedSet but much cheaper O(1) for Dictionary.
Iterate through values then O(n) for SortedSet but expensive O(n* logN)for Dictionary before consumption.

erdembayar commented:
Code reviews please? @zivkan  @nkolev92 @dtivel

nkolev92 commented on src/NuGet.Core/NuGet.ProjectModel/DependencyGraphSpec.cs:
```diff
@@ -18,8 +19,9 @@ public class DependencyGraphSpec
 
-        private readonly SortedSet<string> _restore = new SortedSet<string>(PathUtility.GetStringComparerBasedOnOS());
-        private readonly SortedDictionary<string, PackageSpec> _projects = new SortedDictionary<string, PackageSpec>(PathUtility.GetStringComparerBasedOnOS());
+        // Here I know we're wasting bool value which is extra overhead, but it's most closest builtin thing to HashSet and make it thread safe.
```
I don't think we need the comment. ConcurrenctHashSet not existing is a known thing :)

erdembayar replied:
Removed unneeded comment.

nkolev92 commented on src/NuGet.Core/NuGet.ProjectModel/DependencyGraphSpec.cs:
```diff
@@ -66,7 +68,7 @@ public IReadOnlyList<string> Restore
             get
             {
-                return _restore.ToList();
+                return _restore.Keys.OrderBy(r => r, PathUtility.GetStringComparerBasedOnOS()).ToList();
```
Did you investigate if/how the ordering is used? 
I'm wondering if there's a need for it in the first. 

It certainly won't be a bottleneck, but no point in incurring additional allocations if we don't absolutely need them.

erdembayar replied:
It looks this ordering matters some places:
1st [here](https://github.com/NuGet/NuGet.Client/blob/3501ddedc274ac10d4b135856b7593a6bb8a72f1/src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/NoOpRestoreUtilities.cs#L196), 2nd [here](https://github.com/NuGet/NuGet.Client/blob/3501ddedc274ac10d4b135856b7593a6bb8a72f1/src/NuGet.Core/NuGet.ProjectModel/ProjectLockFile/PackagesLockFileUtilities.cs#L78). So better to keep it, as you mentioned it's not bottleneck.

nkolev92 replied:
#2 is always going to have only 1 project. 
#1 is valid, that affects no-op. 

We can keep it for now, it makes sense.

PS. When it's not asking for a code change, please allow the reviewer to resolve the thread. It took me some time to find the comment when I clicked the link since it was collapsed.

zivkan replied:
never mind. Was going to ask something, but changed my mind.

erdembayar committed:
Remove unneeded comment.

nkolev92 reviewed (commented):
There's a TODO for this issue in the code: 

https://github.com/NuGet/NuGet.Client/blob/935b1d2f5bef8d0d78a737190846b72f0bae30e9/src/NuGet.Core/NuGet.Build.Tasks.Console/MSBuildStaticGraphRestore.cs#L625

Can you please resolve that? 
Ready for approval after that I think.

zivkan reviewed (dismissed):
This PR is less effort than my preferred option, but I'd prefer a more "functional" (less side-effects) approach. In my opinion, DependencyGraphSpec should be immutable, and the parallel foreach should not mutate the object for each package, but should do create the PackageSpec in parallel, then use the orchestrating task/thread to accumulate the results in a lock-free way. Very roughly:

```cs
var tasks = new List<Task<PackageSpec>>();
var packageSpecs = List<PackageSpec>();
IEnumerable<Project> enumerator = projects.GetEnumerable();

for (int i = 0; i < Environment.ProcessorCount; i++)
{
  if (!enumerator.MoveNext())
  {
    break;
  }
  tasks.Add(Task.Run(() => GetProjectSpec(enumerator.Current)));
}

while (tasks.Count > 0)
{
  var finished = await Task.WhenAny(tasks);
  tasks.Remove(finished);
  packageSpecs.Add(finished.Result);

  if (enumerator.Current != null && enumerator.MoveNext())
  {
    tasks.Add(() => GetPackageSpec(enumerator.Current));
  }
}

var dgSpec = new DependencyGraphSpec(packageSpecs);
```

Even better would be to use TPL dataflow, rather than building our own concurrency limiter. In one of Martin's PRs I already provided sample code on how the same thing is achieveable with TPL dataflow. I remember that NuGet.Build.Tasks.Console already uses TPL dataflow, so it should be very easy to use there. Making the DGSpec immutable would be much less easy.

erdembayar committed:
Remove syncronization lock inside MSBuildStaticGraphRestore.cs since we made .AddProject support concurrency.

erdembayar committed:
Merge branch 'dev' into dev-eryondon-enableDGSpecSupportConcurrency2

erdembayar commented:
> There's a TODO for this issue in the code:
> 
> https://github.com/NuGet/NuGet.Client/blob/935b1d2f5bef8d0d78a737190846b72f0bae30e9/src/NuGet.Core/NuGet.Build.Tasks.Console/MSBuildStaticGraphRestore.cs#L625
> 
> Can you please resolve that?
> Ready for approval after that I think.

Ok. Resolved it.

zivkan commented on src/NuGet.Core/NuGet.ProjectModel/DependencyGraphSpec.cs:
```diff
@@ -253,7 +254,7 @@ public void AddProject(PackageSpec projectSpec)
                 }
 
-                _projects.Add(projectUniqueName, projectToRestore);
+                _projects.TryAdd(projectUniqueName, projectToRestore);
```
The semantics here are probably changing. I'm not sure about `SortedDictionary<TKey, TValue>`, but `Dictionary<TKey, TValue>` throws when you try to call `.Add` on a key that already exists. The new code will silently ignore the new value.

zivkan commented on src/NuGet.Core/NuGet.ProjectModel/DependencyGraphSpec.cs:
```diff
@@ -323,7 +324,7 @@ public static DependencyGraphSpec Load(string path)
                                     PackageSpec packageSpec = JsonPackageSpecReader.GetPackageSpec(jsonReader, path);
 
-                                    dgspec._projects.Add(projectsPropertyName, packageSpec);
+                                    dgspec._projects.TryAdd(projectsPropertyName, packageSpec);
```
The semantics here are probably changing. I'm not sure about `SortedDictionary<TKey, TValue>`, but `Dictionary<TKey, TValue>` throws when you try to call `.Add` on a key that already exists. The new code will silently ignore the new value.

erdembayar replied:
Changed the syntax.

erdembayar replied:
I added wrapper function check duplicate key value before add value.

> This PR is less effort than my preferred option, but I'd prefer a more "functional" (less side-effects) approach. In my opinion, DependencyGraphSpec should be immutable, and the parallel foreach should not mutate the object for each package, but should do create the PackageSpec in parallel, then use the orchestrating task/thread to accumulate the results in a lock-free way. Very roughly:
> 
> ```cs
> var tasks = new List<Task<PackageSpec>>();
> var packageSpecs = List<PackageSpec>();
> IEnumerable<Project> enumerator = projects.GetEnumerable();
> 
> for (int i = 0; i < Environment.ProcessorCount; i++)
> {
>   if (!enumerator.MoveNext())
>   {
>     break;
>   }
>   tasks.Add(Task.Run(() => GetProjectSpec(enumerator.Current)));
> }
> 
> while (tasks.Count > 0)
> {
>   var finished = await Task.WhenAny(tasks);
>   tasks.Remove(finished);
>   packageSpecs.Add(finished.Result);
> 
>   if (enumerator.Current != null && enumerator.MoveNext())
>   {
>     tasks.Add(() => GetPackageSpec(enumerator.Current));
>   }
> }
> 
> var dgSpec = new DependencyGraphSpec(packageSpecs);
> ```
> 
> Even better would be to use TPL dataflow, rather than building our own concurrency limiter. In one of Martin's PRs I already provided sample code on how the same thing is achieveable with TPL dataflow. I remember that NuGet.Build.Tasks.Console already uses TPL dataflow, so it should be very easy to use there. Making the DGSpec immutable would be much less easy.

I'll take into my other PR.

zivkan commented on src/NuGet.Core/NuGet.ProjectModel/DependencyGraphSpec.cs:
```diff
@@ -383,7 +385,7 @@ private void ParseJson(JObject json)
                     var spec = JsonPackageSpecReader.GetPackageSpec(specJson);
 #pragma warning restore CS0618
-                    _projects.Add(prop.Name, spec);
+                    _projects.TryAdd(prop.Name, spec);
```
The semantics here are probably changing. I'm not sure about `SortedDictionary<TKey, TValue>`, but `Dictionary<TKey, TValue>` throws when you try to call `.Add` on a key that already exists. The new code will silently ignore the new value.

erdembayar replied:
Changed the syntax.

zivkan replied:
using `dict[key] = value` does not reproduce the semantics of `Dictionary<T1, T2>()`'s `.Add` method which throws an exception when the key already exists.

I have no idea if any of our code depends on that behaviour, throwing if the key already exists, so maybe it's not a problem. But the changed syntax doesn't maintain semantics.

erdembayar replied:
Ok. I added wrapper function `ConcurrentDictAddCheckDuplicateKey` which check for duplicate key before add value and throw exception if it's duplicate key to keep existing behaviour. Please check.

zivkan commented on src/NuGet.Core/NuGet.ProjectModel/DependencyGraphSpec.cs:
```diff
@@ -371,7 +372,8 @@ private void ParseJson(JObject json)
             {
-                _restore.UnionWith(restoreObj.Properties().Select(prop => prop.Name));
+                _restore.AddRange(restoreObj.Properties().Select(prop => prop.Name)
+                    .ToDictionary(name => name, name => false));
```
micro-optimisation: creating a dictionary, just to pass to `AddRange` is a bit of a waste of CPU time. Changing the `Select(prop => prop.Name)` to `Select(props => new KeyValuePair<string, bool>(prop.Name, false)` should have the same result, but without the dictionary needing to hash all the keys, check for duplicates, etc.

erdembayar replied:
Ok. Taken the recommendation.

zivkan reviewed (commented):
Trying to reason about multi-threaded correctness is difficult. It seems to me that `_restore` and `_projects` are never both accessed in the same method, so it seems ok to me. And at the moment is looks like only NuGet.Build.Tasks.Console was trying to use it in parallel, and only to add.

Having said that, I've read enumerating `ConcurrentDictionary` is expensive, so if this class' `Restore` and `Projects` properties are accessed more often than the methods that add things to the dictionary, this PR might reduce performance, not improve it.

erdembayar committed:
Address code review comments by Andy.

erdembayar commented on src/NuGet.Core/NuGet.Build.Tasks.Console/MSBuildStaticGraphRestore.cs:
```diff
@@ -622,11 +622,7 @@ private DependencyGraphSpec GetDependencyGraphSpec(string entryProjectPath, IDic
                                 projectPathLookup.TryAdd(projectPath, projectPath);
                             }
 
-                            // TODO: Remove this lock once https://github.com/NuGet/Home/issues/9002 is fixed
```
@nkolev92  I removed this todo comment with sync lock.

erdembayar committed:
Revert "Address code review comments by Andy."

This reverts commit a6d28988c3e59dfbce98af67104a06c680c25f1e.

donnie-msft commented on src/NuGet.Core/NuGet.ProjectModel/DependencyGraphSpec.cs:
```diff
@@ -18,8 +19,8 @@ public class DependencyGraphSpec
 
-        private readonly SortedSet<string> _restore = new SortedSet<string>(PathUtility.GetStringComparerBasedOnOS());
-        private readonly SortedDictionary<string, PackageSpec> _projects = new SortedDictionary<string, PackageSpec>(PathUtility.GetStringComparerBasedOnOS());
+        private readonly ConcurrentDictionary<string, bool> _restore = new ConcurrentDictionary<string, bool>();
+        private readonly ConcurrentDictionary<string, PackageSpec> _projects = new ConcurrentDictionary<string, PackageSpec>(PathUtility.GetStringComparerBasedOnOS());
```
Why does _restore not use a comparer, while _projects does?

nkolev92 replied:
I think both should use the comparer as the previous implementation did.

erdembayar replied:
Good catch. Fixed.

donnie-msft commented on src/NuGet.Core/NuGet.ProjectModel/DependencyGraphSpec.cs:
```diff
@@ -422,9 +423,9 @@ private void Write(RuntimeModel.IObjectWriter writer, Action<PackageSpec, Runtim
 
             // Preserve default sort order
-            foreach (var restoreName in _restore)
+            foreach (var restoreName in _restore.OrderBy(r => r.Key, PathUtility.GetStringComparerBasedOnOS()))
```
Isn't this already sorted via comparer to `_restore`'s constructor?

erdembayar replied:
No, it's not anymore. Because there is no sorted version of Concurrent Set nor Concurrent Dictionary.  So now insertion is cheaper O(1) instead of O(LogN) but pay more price here O(N LogN). Everything has 2 sides, only depending when how pay the price.

nkolev92 commented:
> This PR is less effort than my preferred option, but I'd prefer a more "functional" (less side-effects) approach. In my opinion, DependencyGraphSpec should be immutable, and the parallel foreach should not mutate the object for each package, but should do create the PackageSpec in parallel, then use the orchestrating task/thread to accumulate the results in a lock-free way. Very roughly:

I don't see that approach as much more effort. 
Thinking about it I do like it better than the proposed/current approach. 

The *only* mutation that we do right now is for CPVM. Do you have any suggestion where to move this mutation? I'm not a fan, but I don't have a suggestion that's not mutating the package spec before adding it to the DGSpec. 

> Having said that, I've read enumerating ConcurrentDictionary is expensive, so if this class' Restore and Projects properties are accessed more often than the methods that add things to the dictionary, this PR might reduce performance, not improve it.

I'm not sure how expensive it'd have to be for this to be true.
It's not necessary a path that repeats thousands of times.

zivkan commented:
> The only mutation that we do right now is for CPVM.

You know the code better than I do, so I want to believe you. But `AddRestore` is references in 11 different files under `src\`, and `AddProject` is referenced by 8 files, so I'm worried that it might not be true.  Having said that, the filenames look like restore entry points, so you're probably right.

> Do you have any suggestion where to move this mutation? I'm not a fan, but I don't have a suggestion that's not mutating the package spec before adding it to the DGSpec.

Well, the theoretical way to do it is for the PackageSpec to also be immutable. When CPVM needs to "modify" a PackageSpec, it creates a new DGSpec with the "to-be-modified" PackageSpec removed, and in its place a new PackageSpec with the required changes.

I don't know if it will be as easy to implement as it was for me to write that though.

We should look into how Roslyn implemented their immutable data types, and see if there's an easy way to create a new instance of a class which is a clone of another instance with just 1 property modified.

zivkan commented:
I assume the "CPVM mutates" thing you're referring to is https://github.com/NuGet/NuGet.Client/blob/d4d80b14ba6ecef2728675070fffc2ae072d84d6/src/NuGet.Core/NuGet.ProjectModel/DependencyGraphSpec.cs#L260-L275 and only that?

Line 262 already does `spec.Clone()`, so maybe it's less mutating that we thought?

Is there anywhere else that you believe CPVM mutates a DGSpec or PackageSpec?

nkolev92 reviewed (approved):
I ok with the current approach assuming the few things we commented on are fixed, but I do like @zivkan's suggestion given that it avoids additional overhead. 

Won't block on it. 

Would be happy to review if necessary.

zivkan commented:
@nkolev92 

I did a real quick benchmark, creating dictionaries with 300 entries to simulate a large solution, and enumerating to only count the number of entries:

|                        Method |      Mean |     Error |    StdDev | Ratio | RatioSD |
|------------------------------ |----------:|----------:|----------:|------:|--------:|
|           EnumerateDictionary |  1.503 us | 0.0057 us | 0.0053 us |  1.00 |    0.00 |
| EnumerateConcurrentDictionary | 10.277 us | 0.0240 us | 0.0201 us |  6.84 |    0.03 |

Note, that with this PR `.Projects` and `.Restore` will also need to sort the result, whereas before this PR these properties used pre-sorted collections, so the impact is potentially greater than what this microbenchmark shows.

Isn't this code used to determine no-op, so this can impact no-op performance?  Or is it only "full" restore?

zivkan commented:
Changing my benchmark program a little to compare enumerating SortedSet, SortedDictionary and ConcurrentDictiory (with LINQ sort), I get these results:

|                        Method |       Mean |     Error |    StdDev | Ratio | RatioSD |
|------------------------------ |-----------:|----------:|----------:|------:|--------:|
|     EnumerateSortedDictionary |   5.730 us | 0.0053 us | 0.0041 us |  1.00 |    0.00 |
| EnumerateConcurrentDictionary | 243.791 us | 2.1654 us | 2.0255 us | 42.62 |    0.34 |
|            EnumerateSortedSet |   5.737 us | 0.0252 us | 0.0235 us |  1.00 |    0.00 |

so, each time it's about 0.24 ms slower to read. Since DGSpec has two properties `Projects` and `Restore`, if they're both accessed just once each, that's only 0.5ms slower. That's probably ok for no-op.  The risk is if they're called more than once per restore.

erdembayar committed:
Address latest code review. No need to check duplicate key for _restore because it was SortedSet. It doesn't throw exception for duplicate value and after converting to ConcurrentDictionary we only care key not value in ConcurrentDictionary.
But for _projects we need to check for duplicate key also care for value in it.

nkolev92 commented:
Without digging, the enumeration will be called at least 3*n number of times where n is the number of projects. (original generation, project specific generation, no-op hash).

We start with a solution dg spec and then we move to a project level dg specs and those are enumerated once or twice during an actual no-op restore.

nkolev92 commented:
so the current experience is somehow worse than the old one. 

@erdembayar had a different proposal where he was using a lock object, so the lock object is likely faster than this. 

Finally, the lockless approach is likely the cleanest if possible.

erdembayar commented:
> @nkolev92
> 
> I did a real quick benchmark, creating dictionaries with 300 entries to simulate a large solution, and enumerating to only count the number of entries:
> 
> Method	Mean	Error	StdDev	Ratio	RatioSD
> EnumerateDictionary	1.503 us	0.0057 us	0.0053 us	1.00	0.00
> EnumerateConcurrentDictionary	10.277 us	0.0240 us	0.0201 us	6.84	0.03
> Note, that with this PR `.Projects` and `.Restore` will also need to sort the result, whereas before this PR these properties used pre-sorted collections.
> 
> Isn't this code used to determine no-op, so this can impact no-op performance? Or is it only "full" restore?

For all sets less than 1k (Solution with 1k projects) usually it's just difference of some nano seconds and even cost of sorting is negligible. Even for SortedSet we used to pay insertion O(LogN) fee but now it's O(1). But paying higher fee before consume by sorting O(N*LogN) instead of O(N).  Much of it cancel/compensate each other.
Also in end result: already reduced restore from 35seconds to 25 seconds on follow up PR.

zivkan commented on src/NuGet.Core/NuGet.ProjectModel/DependencyGraphSpec.cs:
```diff
@@ -527,5 +531,15 @@ private static string[] GetPackageSpecDependencyIds(PackageSpec spec)
+
+        private static void ConcurrentDictAddCheckDuplicateKey<K, V>(ConcurrentDictionary<K, V> dict, K key, V value)
+        {
+            if (dict.ContainsKey(key))
+            {
+                throw new ArgumentException("An element with the same key already exists in the System.Collections.Generic.Dictionary: " + nameof(dict) + " key:" + key);
+            }
+
+            dict.TryAdd(key, value);
```
This looks up the key twice, making the add twice as slow, plus introducing a multi-threaded timing bug (think about when two threads try to add the same key at the same time)

```suggestion
            if (!dict.TryAdd(key, value))
            {
                throw new ArgumentException("An element with the same key already exists in the System.Collections.Generic.Dictionary: " + nameof(dict) + " key:" + key);
            }
```

erdembayar replied:
I see. I can just use synchronization lock. @zivkan please check this [PR#3594](https://github.com/NuGet/NuGet.Client/pull/3594). Maybe other one is much easy to understand.

erdembayar replied:
It seems we can't avoid sync lock, to fix this I need to add sync lock. This PR getting bit complicated maybe just other one is fine. PR#3594.

zivkan replied:
In this method, you can avoid a lock by using the suggestion I made. The dictionary method is `TryAdd`, hence it already tells you if it's successful or not. You don't need a lock.

However, I 100% agree that this PR is problematic. It's trading off read performance for concurrent add. I don't think that's a good trade off.

erdembayar commented:
> Changing my benchmark program a little to compare enumerating SortedSet, SortedDictionary and ConcurrentDictiory (with LINQ sort), I get these results:
> 
> Method	Mean	Error	StdDev	Ratio	RatioSD
> EnumerateSortedDictionary	5.730 us	0.0053 us	0.0041 us	1.00	0.00
> EnumerateConcurrentDictionary	243.791 us	2.1654 us	2.0255 us	42.62	0.34
> EnumerateSortedSet	5.737 us	0.0252 us	0.0235 us	1.00	0.00
> so, each time it's about 0.24 ms slower to read. Since DGSpec has two properties `Projects` and `Restore`, if they're both accessed just once each, that's only 0.5ms slower. That's probably ok for no-op. The risk is if they're called more than once per restore.

I can just use synchronization lock. @zivkan please check this [PR#3594](https://github.com/NuGet/NuGet.Client/pull/3594).

zivkan commented on src/NuGet.Core/NuGet.ProjectModel/DependencyGraphSpec.cs:
```diff
@@ -527,5 +531,15 @@ private static string[] GetPackageSpecDependencyIds(PackageSpec spec)
+        {
+            if (dict.ContainsKey(key))
+            {
+                throw new ArgumentException("An element with the same key already exists in the System.Collections.Generic.Dictionary: " + nameof(dict) + " key:" + key);
```
`dict` is the parameter name of the private method, so putting that name in the exception string doesn't help callers of the public API.

erdembayar replied:
@zivkan I can just use synchronization lock. @zivkan please check this [PR#3594](https://github.com/NuGet/NuGet.Client/pull/3594).

erdembayar commented:
> Changing my benchmark program a little to compare enumerating SortedSet, SortedDictionary and ConcurrentDictiory (with LINQ sort), I get these results:
> 
> Method	Mean	Error	StdDev	Ratio	RatioSD
> EnumerateSortedDictionary	5.730 us	0.0053 us	0.0041 us	1.00	0.00
> EnumerateConcurrentDictionary	243.791 us	2.1654 us	2.0255 us	42.62	0.34
> EnumerateSortedSet	5.737 us	0.0252 us	0.0235 us	1.00	0.00
> so, each time it's about 0.24 ms slower to read. Since DGSpec has two properties `Projects` and `Restore`, if they're both accessed just once each, that's only 0.5ms slower. That's probably ok for no-op. The risk is if they're called more than once per restore.

Can you share your snippet?

erdembayar commented:
> so the current experience is somehow worse than the old one.
> 
> @erdembayar had a different proposal where he was using a lock object, so the lock object is likely faster than this.
> 
> Finally, the lockless approach is likely the cleanest if possible.

I can just use synchronization lock. @zivkan please check this [PR#3594](https://github.com/NuGet/NuGet.Client/pull/3594).

zivkan commented:
> Can you share your snippet?

```cs
public class DirectoryEnumerationBenchmarks
{
    private SortedDictionary<string, string> _sortedDict;
    private ConcurrentDictionary<string, string> _concurrentDict;
    private SortedSet<string> _sortedSet;

    public DirectoryEnumerationBenchmarks()
    {
        _sortedDict = new SortedDictionary<string, string>();
        _concurrentDict = new ConcurrentDictionary<string, string>();
        _sortedSet = new SortedSet<string>();

        for (int i = 300; i > 0; i--)
        {
            var val = $"Project{i}";
            _sortedDict.Add(val, val);
            _concurrentDict.TryAdd(val, val);
            _sortedSet.Add(val);
        }
    }

    [Benchmark(Baseline = true)]
    public int EnumerateSortedDictionary()
    {
        int count = 0;
        foreach (var kvp in _sortedDict)
        {
            count++;
        }
        return count;
    }

    [Benchmark]
    public int EnumerateConcurrentDictionary()
    {
        int count = 0;
        foreach (var kvp in _concurrentDict.OrderBy(k => k.Key))
        {
            count++;
        }
        return count;
    }

    [Benchmark]
    public int EnumerateSortedSet()
    {
        int count = 0;
        foreach (var key in _sortedSet)
        {
            count++;
        }
        return count;
    }
}
```

zivkan commented:
> I can just use synchronization lock.

Although this PR seemed nicer at first, as we get to better understand how it all works, it seems that the `Project` and `Restore` property enumeration performance tradeoffs are not worthwhile, so this PR's approach might not be a good option. I'll start some new comments in the other PR, but my recommendation would be to close this PR, as it's too risky. We'd need to do real-world perf benchmarks to feel confident in accepting this PR.
